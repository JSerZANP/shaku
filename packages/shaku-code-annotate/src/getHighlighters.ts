import { Highlighter, HighlighterOptions, getHighlighter } from "shiki";
import { codeToShakuHtml } from "./codeToShakuHtml";

interface ShakuHighlighter extends Highlighter {
  codeToShakuHtml: typeof codeToShakuHtml;
}

const shikiHighlighterFetcherStore = new Map<
  string,
  Promise<Array<ShakuHighlighter>>
>();

export async function getShakuHighlighters(options: HighlighterOptions) {
  const key = JSON.stringify(options);
  if (shikiHighlighterFetcherStore.has(key)) {
    return shikiHighlighterFetcherStore.get(key)!;
  }

  const themes = (options == null ? void 0 : options.themes) ?? [
    (options == null ? void 0 : options.theme) ?? "github-light",
  ];

  const highlighterFetcher = Promise.all(
    themes.map((theme) =>
      getHighlighter({
        ...(options ?? {}),
        theme,
        themes: void 0,
      }).then((highlighter) => {
        return Object.assign(highlighter, {
          codeToShakuHtml,
        });
      })
    )
  );
  shikiHighlighterFetcherStore.set(key, highlighterFetcher);
  return highlighterFetcher;
}
