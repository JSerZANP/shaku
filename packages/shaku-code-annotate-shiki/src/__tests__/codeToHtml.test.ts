import { expect, test } from "vitest";
import { getShakuHighlighters } from "../getHighlighters";

const snippets = [
  `
function ChatRoom({ roomId }) {
  const [serverUrl, setServerUrl] = useState('https://localhost:1234');
  //       ^
  // [<a href="https:jser.dev">jser.dev</a>]

  useEffect(() => {
     // (                                                     )
        const connection = createConnection(serverUrl, roomId);
     // (1                )
        connection.connect();
    
  	return () => {
           //(2                    )
             connection.disconnect();
  	};
   //(3                 )
  }, [serverUrl, roomId]);
}
`,
  `
function hello() {
  //        ^!
  //    [Hello world!]!
  //        ^
  //    [Shaku <b>Shaku</b>]
  const blog = "https://jser.dev"
}
  `,
  `
function useSomeEffect({blog}) {
  useEffect(() => {
    //  do some stuff
      // @fold start
      return () => {
        // @highlight
        location.href = 'https://jser.dev'
      }
      // @fold ^
    }, [blog])
}
  `,

  `
// @class abc h123
// @highlight
function useSomeEffect({blog}) {
  useEffect(() => {
      return () => {
        location.href = 'https://jser.dev'
      }
    }, [blog])
}
  `,
];

test("codeToHtml()", async () => {
  const highlighters = await getShakuHighlighters({
    // @ts-ignore
    langs: ["js"],
  });

  snippets.forEach((code) => {
    const html = highlighters[0].codeToShakuHtml({
      code,
      options: {
        lang: "js",
      },
    });

    expect(html).toMatchSnapshot();
  });
});

test("codeToHtml() + raw HTML", async () => {
  const highlighters = await getShakuHighlighters({
    // @ts-ignore
    langs: ["js"],
  });

  snippets.forEach((code) => {
    const html = highlighters[0].codeToShakuHtml({
      code,
      options: {
        lang: "js",
        useDangerousRawHtml: true,
      },
    });

    expect(html).toMatchSnapshot();
  });
});
