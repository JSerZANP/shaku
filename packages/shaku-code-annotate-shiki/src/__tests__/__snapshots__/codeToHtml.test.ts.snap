// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codeToHtml() + raw HTML 1`] = `
{
  "html": "<pre class=\\"shiki shaku github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">function</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">ChatRoom</span><span style=\\"color: #24292E\\">({ </span><span style=\\"color: #E36209\\">roomId</span><span style=\\"color: #24292E\\"> }) {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> [</span><span style=\\"color: #005CC5\\">serverUrl</span><span style=\\"color: #24292E\\">, </span><span style=\\"color: #005CC5\\">setServerUrl</span><span style=\\"color: #24292E\\">] </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">useState</span><span style=\\"color: #24292E\\">(</span><span style=\\"color: #032F62\\">&#039;https://localhost:1234&#039;</span><span style=\\"color: #24292E\\">);</span></div><div class=\\"shaku-callout\\" style=\\"left:5ch\\"><span class=\\"shaku-callout-arrow\\" style=\\"left:6ch\\"></span><a href=\\"https:jser.dev\\">jser.dev</a></div><div class=\\"line\\"></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #6F42C1\\">useEffect</span><span style=\\"color: #24292E\\">(() </span><span style=\\"color: #D73A49\\">=&gt;</span><span style=\\"color: #24292E\\"> {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">        </span><mark class=\\"shaku-inline-highlight\\" ><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">connection</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">createConnection</span><span style=\\"color: #24292E\\">(serverUrl, roomId);</span></mark></div><div class=\\"line\\"><span style=\\"color: #24292E\\">        </span><mark class=\\"shaku-inline-highlight\\" data-id=\\"1\\"><span style=\\"color: #24292E\\">connection.</span><span style=\\"color: #6F42C1\\">connect</span><span style=\\"color: #24292E\\">(</span></mark><span style=\\"color: #24292E\\">);</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">    </span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  	</span><span style=\\"color: #D73A49\\">return</span><span style=\\"color: #24292E\\"> () </span><span style=\\"color: #D73A49\\">=&gt;</span><span style=\\"color: #24292E\\"> {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">             </span><mark class=\\"shaku-inline-highlight\\" data-id=\\"2\\"><span style=\\"color: #24292E\\">connection.</span><span style=\\"color: #6F42C1\\">disconnect</span><span style=\\"color: #24292E\\">()</span></mark><span style=\\"color: #24292E\\">;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  	};</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  }, </span><mark class=\\"shaku-inline-highlight\\" data-id=\\"3\\"><span style=\\"color: #24292E\\">[serverUrl, roomId])</span></mark><span style=\\"color: #24292E\\">;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">}</span></div><div class=\\"line\\"></div></code></div></pre>",
  "skipped": false,
}
`;

exports[`codeToHtml() + raw HTML 2`] = `
{
  "html": "<pre class=\\"shiki shaku github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">function</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">hello</span><span style=\\"color: #24292E\\">() {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #6A737D\\">//        ^</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #6A737D\\">//    [Hello world!]</span></div><div class=\\"shaku-callout\\" style=\\"left:8ch\\"><span class=\\"shaku-callout-arrow\\" style=\\"left:4ch\\"></span>Shaku <b>Shaku</b></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">blog</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #032F62\\">&quot;https://jser.dev&quot;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">}</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span></div></code></div></pre>",
  "skipped": false,
}
`;

exports[`codeToHtml() + raw HTML 3`] = `
{
  "html": "<pre class=\\"shiki shaku github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">function</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">useSomeEffect</span><span style=\\"color: #24292E\\">({</span><span style=\\"color: #E36209\\">blog</span><span style=\\"color: #24292E\\">}) {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #6F42C1\\">useEffect</span><span style=\\"color: #24292E\\">(() </span><span style=\\"color: #D73A49\\">=&gt;</span><span style=\\"color: #24292E\\"> {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">    </span><span style=\\"color: #6A737D\\">//  do some stuff</span></div><details class=\\"shaku-expand\\"><summary style=\\"margin-left:6ch\\"><mark>{...}</mark></summary><div class=\\"line\\"><span style=\\"color: #24292E\\">      </span><span style=\\"color: #D73A49\\">return</span><span style=\\"color: #24292E\\"> () </span><span style=\\"color: #D73A49\\">=&gt;</span><span style=\\"color: #24292E\\"> {</span></div><div class=\\"line highlight\\"><span style=\\"color: #24292E\\">        location.href </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #032F62\\">&#039;https://jser.dev&#039;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">      }</span></div></details><div class=\\"line\\"><span style=\\"color: #24292E\\">    }, [blog])</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">}</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span></div></code></div></pre>",
  "skipped": false,
}
`;

exports[`codeToHtml() 1`] = `
{
  "html": "<pre class=\\"shiki shaku github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">function</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">ChatRoom</span><span style=\\"color: #24292E\\">({ </span><span style=\\"color: #E36209\\">roomId</span><span style=\\"color: #24292E\\"> }) {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> [</span><span style=\\"color: #005CC5\\">serverUrl</span><span style=\\"color: #24292E\\">, </span><span style=\\"color: #005CC5\\">setServerUrl</span><span style=\\"color: #24292E\\">] </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">useState</span><span style=\\"color: #24292E\\">(</span><span style=\\"color: #032F62\\">&#039;https://localhost:1234&#039;</span><span style=\\"color: #24292E\\">);</span></div><div class=\\"shaku-callout\\" style=\\"left:5ch\\"><span class=\\"shaku-callout-arrow\\" style=\\"left:6ch\\"></span>&lt;a href=&quot;https:jser.dev&quot;&gt;jser.dev&lt;/a&gt;</div><div class=\\"line\\"></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #6F42C1\\">useEffect</span><span style=\\"color: #24292E\\">(() </span><span style=\\"color: #D73A49\\">=&gt;</span><span style=\\"color: #24292E\\"> {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">        </span><mark class=\\"shaku-inline-highlight\\" ><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">connection</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">createConnection</span><span style=\\"color: #24292E\\">(serverUrl, roomId);</span></mark></div><div class=\\"line\\"><span style=\\"color: #24292E\\">        </span><mark class=\\"shaku-inline-highlight\\" data-id=\\"1\\"><span style=\\"color: #24292E\\">connection.</span><span style=\\"color: #6F42C1\\">connect</span><span style=\\"color: #24292E\\">(</span></mark><span style=\\"color: #24292E\\">);</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">    </span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  	</span><span style=\\"color: #D73A49\\">return</span><span style=\\"color: #24292E\\"> () </span><span style=\\"color: #D73A49\\">=&gt;</span><span style=\\"color: #24292E\\"> {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">             </span><mark class=\\"shaku-inline-highlight\\" data-id=\\"2\\"><span style=\\"color: #24292E\\">connection.</span><span style=\\"color: #6F42C1\\">disconnect</span><span style=\\"color: #24292E\\">()</span></mark><span style=\\"color: #24292E\\">;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  	};</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  }, </span><mark class=\\"shaku-inline-highlight\\" data-id=\\"3\\"><span style=\\"color: #24292E\\">[serverUrl, roomId])</span></mark><span style=\\"color: #24292E\\">;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">}</span></div><div class=\\"line\\"></div></code></div></pre>",
  "skipped": false,
}
`;

exports[`codeToHtml() 2`] = `
{
  "html": "<pre class=\\"shiki shaku github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">function</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">hello</span><span style=\\"color: #24292E\\">() {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #6A737D\\">//        ^</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #6A737D\\">//    [Hello world!]</span></div><div class=\\"shaku-callout\\" style=\\"left:8ch\\"><span class=\\"shaku-callout-arrow\\" style=\\"left:4ch\\"></span>Shaku &lt;b&gt;Shaku&lt;/b&gt;</div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #D73A49\\">const</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #005CC5\\">blog</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #032F62\\">&quot;https://jser.dev&quot;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">}</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span></div></code></div></pre>",
  "skipped": false,
}
`;

exports[`codeToHtml() 3`] = `
{
  "html": "<pre class=\\"shiki shaku github-light\\" style=\\"color:#24292e;background-color:#fff\\"><div class=\\"code-container\\"><code><div class=\\"line\\"></div><div class=\\"line\\"><span style=\\"color: #D73A49\\">function</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #6F42C1\\">useSomeEffect</span><span style=\\"color: #24292E\\">({</span><span style=\\"color: #E36209\\">blog</span><span style=\\"color: #24292E\\">}) {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span><span style=\\"color: #6F42C1\\">useEffect</span><span style=\\"color: #24292E\\">(() </span><span style=\\"color: #D73A49\\">=&gt;</span><span style=\\"color: #24292E\\"> {</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">    </span><span style=\\"color: #6A737D\\">//  do some stuff</span></div><details class=\\"shaku-expand\\"><summary style=\\"margin-left:6ch\\"><mark>{...}</mark></summary><div class=\\"line\\"><span style=\\"color: #24292E\\">      </span><span style=\\"color: #D73A49\\">return</span><span style=\\"color: #24292E\\"> () </span><span style=\\"color: #D73A49\\">=&gt;</span><span style=\\"color: #24292E\\"> {</span></div><div class=\\"line highlight\\"><span style=\\"color: #24292E\\">        location.href </span><span style=\\"color: #D73A49\\">=</span><span style=\\"color: #24292E\\"> </span><span style=\\"color: #032F62\\">&#039;https://jser.dev&#039;</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">      }</span></div></details><div class=\\"line\\"><span style=\\"color: #24292E\\">    }, [blog])</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">}</span></div><div class=\\"line\\"><span style=\\"color: #24292E\\">  </span></div></code></div></pre>",
  "skipped": false,
}
`;
